### PR TITLE
feat: support contextFormatter

### PR DESCRIPTION
**Pull request by popomore**
##### Checklist

- [ ] `npm test` passes
- [ ] tests and/or benchmarks are included
- [ ] documentation is changed or added
- [ ] commit message follows commit guidelines

##### Description of change

Ref https://github.com/eggjs/egg-logger/pull/46 https://github.com/eggjs/egg-logger/pull/37

### CODE REVIEW

File: lib/utils.js [hunk 1]
```diff
@@ -80,7 +80,8 @@ module.exports = {
     meta = meta || {};
     let message;
     let output;
-    const formatter = meta.formatter || options.formatter;
+    let formatter = meta.formatter || options.formatter;
+    if (meta.ctx && options.contextFormatter) formatter = options.contextFormatter;
```
Comment (popomore) [inline lib/utils.js]: contextFormatter 没有默认值，如果定义了则用之

File: index.d.ts [hunk 2]
```diff
@@ -20,6 +20,7 @@ export interface LoggerOptions {
 export interface EggLoggerOptions extends LoggerOptions {
   file: string;
   formatter?: any;
```
Comment (atian25) [inline index.d.ts]:
```suggestion
  formatter?: (meta?: object) => string;
  contextFormatter?: (meta?: object) => string;
```

File: index.d.ts [hunk 3]
```diff
@@ -20,6 +20,7 @@ export interface LoggerOptions {
...
+  contextFormatter?: any;
```
Comment (atian25) [inline index.d.ts]: contextFormatter?: (meta?: object) => string;

Comment (atian25) [inline index.d.ts]:
```suggestion
  contextFormatter?: (meta?: object) => string;
```

Comment (atian25) [inline index.d.ts]:
```suggestion
  contextFormatter?: (meta?: object) => string;
```

File: lib/egg/context_logger.js [hunk 4]
```diff
@@ -44,6 +44,7 @@ class ContextLogger {
     const meta = {
       formatter: contextFormatter,
       paddingMessage: this.paddingMessage,
+      ctx: this.ctx,
```
Comment (atian25) [inline lib/egg/context_logger.js]: 要不要 define 为不可枚举的，避免之前的 logger 有 JSON.stringify 的

**Comment by atian25** [comment]
@popomore 顺便把依赖换下？

https://github.com/eggjs/egg/issues/3530

> warning egg > egg-cluster > egg-logger > circular-json@0.5.9: CircularJSON is in maintenance only, flatted is its successor.

换为 circular-json-for-egg

**Comment by popomore** [comment]
+ egg-logger@2.4.0

**Pull request merged**